src/interface.jl

```diff
@@ -2,86 +2,185 @@
 
 
 """
-    hasdensity(d)::Bool
+    abstract type DensityKind end
 
-Return `true` if `d` is compatible with the `DensityInterface` interface.
+Abstract supertype for types returned by `densitykind(object)`.
+"""
+abstract type DensityKind end
+export DensityKind
+
+"""
+    abstract type HasNoDensity end
+
+Possible return value of `densitykind(object)`. Indicates that
+`object` *is not and does not have* a density, as understood by `DensityInterface`.
+"""
+abstract type HasNoDensity <: DensityKind end
+export HasNoDensity
+
+"""
+    abstract type IsOrHasDensity end
+
+Possible return value of `densitykind(object)`. Indicates that
+`object` either *is* or *has* a density, as understood by `DensityInterface`.
+
+See also [`logdensityof`](@ref) and [`densityof`](@ref).
+"""
+abstract type IsOrHasDensity <: DensityKind end
+export IsOrHasDensity
 
-`hasdensity(d) == true` implies that `d` is either a density itself or has an
-associated density, e.g. a probability density function or a Radon–Nikodym
-derivative with respect to an implicit base measure. It also implies that the
-value of that density at given points can be calculated via
+"""
+    abstract type IsDensity end
+
+Possible return value of `densitykind(object)`. Indicates that
+`object` *is* a density, as understood by `DensityInterface`.
+
+See also [`logdensityof`](@ref) and [`densityof`](@ref).
+"""
+abstract type IsDensity <: IsOrHasDensity end
+export IsDensity
+
+
+"""
+    densitykind(object)::Type{<:DensityKind}
+
+Return a subtype of `IsOrHasDensity` or `HasNoDensity`, to indicate if `object` is
+compatible with the `DensityInterface` API.
+
+`densitykind(object) <: IsOrHasDensity` implies that `object` is either a density
+itself or has an associated density, e.g. a probability density function or a
+Radon–Nikodym derivative with respect to an implicit base measure. It also
+implies that the value of that density at given points can be calculated via
 [`logdensityof`](@ref) and [`densityof`](@ref).
+
+The subtype of [`IsOrHasDensity`](@ref) provides additional information on the
+relationship between the object and the density (i.e. whether `object`
+itself *is* a density or in what way it *has* a density).
+
+Defaults to `HasNoDensity`. For types that directly represent densities, define
+
+```julia
+@inline densitykind(::MyDensityType) = IsDensity
 ```
+
+For other types that are associated with a densitiy (e.g. measure types) use
+other appropriate subtypes (defined outside of `DensityInterface`) of
+`IsOrHasDensity` (instead of `IsDensity`).
 """
-function hasdensity end
-export hasdensity
+function densitykind end
+export densitykind
 
-@inline hasdensity(::Any) = false
+@inline densitykind(object) = HasNoDensity
```

Comment:
Wait a minute... I think the more usual thing is to return an instance here (the empty structs would be singletons). At least `IteratorSize` does that, too. Is there any good reason to make the density kinds abstract types, except for the `IsOrHasDensity` intermediate?

Reply:
Also, it makes dispatch harder. You always have to write `::Type{IsDensity}` instead of `::IsDensity` in trait methods.

Reply:
Yes, we could makes the "leaf type" singleton structs. Would prevent packages to introduce custom subtypes of `IsDensity` though, that's why I chose to go with abstract types for all.

Reply:
Returning an instance is preferable to returning the type.

Reply:
But why? We can do it, but it would make it impossible to "sub-class" concrete density "kinds".

Reply:
Ok, I changed it, `densitykind(object)` returns (singleon) instances now, not types.